net/mptcp/mptcp_sched.c

```diff
@@ -138,53 +240,25 @@ static struct sock *get_available_subflow(struct sock *meta_sk,
 		}
 	}
 
-	/* First, find the best subflow */
-	mptcp_for_each_sk(mpcb, sk) {
-		struct tcp_sock *tp = tcp_sk(sk);
-
-		if (tp->mptcp->rcv_low_prio || tp->mptcp->low_prio)
-			cnt_backups++;
-
-		if ((tp->mptcp->rcv_low_prio || tp->mptcp->low_prio) &&
-		    tp->srtt_us < lowprio_min_time_to_peer) {
-			if (!mptcp_is_available(sk, skb, zero_wnd_test))
-				continue;
-
-			if (mptcp_dont_reinject_skb(tp, skb)) {
-				backupsk = sk;
-				continue;
-			}
-
-			lowprio_min_time_to_peer = tp->srtt_us;
-			lowpriosk = sk;
-		} else if (!(tp->mptcp->rcv_low_prio || tp->mptcp->low_prio) &&
-			   tp->srtt_us < min_time_to_peer) {
-			if (!mptcp_is_available(sk, skb, zero_wnd_test))
-				continue;
-
-			if (mptcp_dont_reinject_skb(tp, skb)) {
-				backupsk = sk;
-				continue;
-			}
-
-			min_time_to_peer = tp->srtt_us;
-			bestsk = sk;
-		}
-	}
-
-	if (mpcb->cnt_established == cnt_backups && lowpriosk) {
-		sk = lowpriosk;
-	} else if (bestsk) {
-		sk = bestsk;
-	} else if (backupsk) {
-		/* It has been sent on all subflows once - let's give it a
-		 * chance again by restarting its pathmask.
+	/* Find the best subflow */
+	sk = get_subflow_from_selectors(mpcb, skb, &subflow_is_active,
+					zero_wnd_test, &force);
+	if (force)
+		/* one unused active sk or one NULL sk when there is at least
+		 * one temporally unavailable unused active sk
 		 */
-		if (skb)
-			TCP_SKB_CB(skb)->path_mask = 0;
-		sk = backupsk;
-	}
+		return sk;
 
+	sk = get_subflow_from_selectors(mpcb, skb, &subflow_is_backup,
+					zero_wnd_test, &force);
+	if (!force)
+		/* one used backup sk or one NULL sk where there is no one
+		 * temporally unavailable unused backup sk
+		 *
+		 * the skb passed through all the available active and backups
+		 * sks, so clean the path mask
+		 */
+		TCP_SKB_CB(skb)->path_mask = 0;
```

Comment:
Improvement: we only clear the path_mask when all the active and backup subflows have been used